File: generator/src/main/java/ch/bailu/gtk/model/NamespaceModel.java
package ch.bailu.gtk.model;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import ch.bailu.gtk.Configuration;
import ch.bailu.gtk.table.NamespaceTable;
import ch.bailu.gtk.converter.RelativeNamespaceType;
import ch.bailu.gtk.tag.NamedWithDocTag;
import ch.bailu.gtk.tag.NamespaceTag;


public class NamespaceModel extends Model implements NamespaceInterface {

    private List<String> includes = new ArrayList<>();

    private final String namespace;


    public NamespaceModel(NamespaceTag namespace) {
        this.namespace = namespace.getName().toLowerCase();

        for (NamedWithDocTag i : namespace.getIncludes()) {
            includes.add(i.getName());
        }

        setSupported("Namespace", NamespaceTable.instance().contains(this.namespace));
    }

    public NamespaceModel() {
        this.namespace = "";
        setSupported("",true);
    }


    public NamespaceModel(RelativeNamespaceType type) {
        namespace = type.getNamespace();
        setSupported("Namespace", NamespaceTable.instance().contains(this.namespace));
    }


    public File getJavaSourceDirectory() {
        File result = new File(Configuration.instance().getJavaBaseDir(), namespace);
        return result;
    }

    public File getCSourceDirectory() {
        return new File(Configuration.instance().getCBaseDir());
    }

    public List<String> getIncludes() {
        return includes;
    }


    @Override
    public String getNamespace() {
        return namespace;
    }

    @Override
    public String getFullNamespace() {
        if ("".equals(namespace)) {
            return Configuration.BASE_NAME_SPACE_NODOT;
        } else {
            return Configuration.BASE_NAME_SPACE_DOT + namespace;
        }
    }
}

File: generator/src/main/kotlin/ch/bailu/gtk/model/NamespaceModel.kt
package ch.bailu.gtk.model

import ch.bailu.gtk.Configuration
import ch.bailu.gtk.NamespaceConfig
import ch.bailu.gtk.model.type.NamespaceType
import ch.bailu.gtk.parser.tag.MethodTag
import ch.bailu.gtk.parser.tag.NamespaceTag
import ch.bailu.gtk.parser.tag.ParameterTag
import ch.bailu.gtk.table.NamespaceTable

class NamespaceModel(
    val namespace: String = "",
    val library: String = "",
    val functions: List<MethodTag> = ArrayList(),
    val constants: List<ParameterTag> = ArrayList(),
    val infoDocUrl: String = "",
    val infoGirFile: String = "",
    val infoNamespace: String = "",
    val infoVersion: String = "",
    val infoSharedLibrary: String = ""
) : Model() {


    constructor(tag: NamespaceTag, config: NamespaceConfig) :
            this(
                namespace = tag.getName().lowercase(),
                library = config.library,
                functions = tag.getFunctions(),
                constants = tag.getConstants(),
                infoDocUrl = config.docUrl.getBaseUrl(),
                infoVersion = tag.version,
                infoNamespace = tag.getName(),
                infoGirFile = config.girFile,
                infoSharedLibrary = tag.sharedLibrary
            ) {
        setSupported("unknown-namespace", NamespaceTable.contains(namespace))
    }

    constructor(type: NamespaceType) : this(type.namespace) {
        setSupported("unknown-namespace", NamespaceTable.contains(namespace))
    }

    val fullNamespace: String
        get() = if ("" == namespace) {
            Configuration.BASE_NAME_SPACE_NODOT
        } else {
            Configuration.BASE_NAME_SPACE_DOT + namespace
        }

}
